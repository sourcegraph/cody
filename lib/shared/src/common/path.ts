import type { URI } from 'vscode-uri'

import { isWindows as _isWindows, isMacOS } from './platform'

interface PathFunctions {
    /**
     * All but the last element of path, or "." if that would be the empty path.
     */
    dirname: (path: string) => string

    /**
     * The last element of path, or "" if path is empty.
     * @param path the path to operate on
     * @param suffix optional suffix to remove
     */
    basename: (path: string, suffix?: string) => string

    /** The extension of path, including the last '.'. */
    extname: (path: string) => string

    /**
     * The relative path from {@link from} to {@link to}.
     */
    relative: (from: string, to: string) => string

    resolve: (...paths: string[]) => string

    /** Path separator. */
    separator: PathSeparator
}

/** For file system paths on Windows ('\' separators, drive letters, case-insensitive). */
export const windowsFilePaths: PathFunctions = pathFunctions(true, '\\', false)

/** For POSIX file system paths ('/' separators, case-sensitive). */
export const posixFilePaths: PathFunctions = pathFunctions(false, '/', true)

/**
 * Get the {@link PathFunctions} to use for the given URI's path ('/' separators, drive letters/case-sensitivity depend on `isWindows`).
 */
export function pathFunctionsForURI(uri: URI, isWindows = _isWindows()): PathFunctions {
    // URIs are always forward slashes even on Windows.
    const sep = '/'
    return uri.scheme === 'file' && isWindows
        ? // Like windowsFilePaths but with forward slashes.
          pathFunctions(true, sep, false)
        : posixFilePaths
}

export function defaultPathFunctions(): PathFunctions {
    const isWindows = _isWindows()
    return pathFunctions(isWindows, isWindows ? '\\' : '/', isMacOS())
}

// I don't like reimplementing this here, but it's the best option because: (1) using Node's `path`
// module requires us to configure the bundler right or risk a subtle bug (and path-browserify
// doesn't have `path/win32` support, so we'd need multiple underlying packages); and (2)
// `vscode-uri` is hard to test because it has a global constant for the current platform set at
// init time.
//
// note: case-insensitive for Windows is not strictly sound, it's possible to have case-sensitive
// volumes on Windows (and case-insensitive on other platforms), however it's impossible to tell from
// a URI/path alone whether that's the case and VS Code already assumes based on platform, so things
// are unfortunately already broken if you don't use the "defaults".
// - https://github.com/microsoft/vscode/issues/94307
// - https://github.com/microsoft/vscode/issues/177925
function pathFunctions(isWindows: boolean, sep: '\\' | '/', caseSensitive: boolean): PathFunctions {
    const f: PathFunctions = {
        dirname(path: string): string {
            if (path === '') {
                return '.'
            }
            if (isWindows && isDriveLetter(path)) {
                return path
            }
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            if (path === '') {
                return sep
            }
            const i = path.lastIndexOf(sep)
            if (i === -1) {
                return '.'
            }
            if (i === 0) {
                return sep
            }
            path = path.slice(0, i)
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            return path
        },
        basename(path: string, suffix?: string): string {
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return ''
            }
            path = path.split(sep).at(-1) ?? ''
            if (suffix && path.endsWith(suffix)) {
                path = path.slice(0, -suffix.length)
            }
            return path
        },
        extname(path: string): string {
            const basename = f.basename(path)
            const i = basename.lastIndexOf('.')
            if (i === 0 || i === -1) {
                return ''
            }
            return basename.slice(i)
        },
        relative(from, to) {
            // For relative paths, output separator is always based on platform, even
            // if the input (sep) is forward slash/URI.
            return relative(from, to, sep, isWindows ? '\\' : '/', caseSensitive)
        },
        resolve(...paths) {
            return paths.join(sep)
        },
        separator: sep,
    }
    return f
}

function isDriveLetter(path: string): boolean {
    return /^[A-Za-z]:$/.test(path)
}

/** The relative path from {@link from} to {@link to}. */
function relative(
    from: string,
    to: string,
    inputSeparator: PathSeparator,
    outputSeparator: PathSeparator,
    caseSensitive: boolean
): string {
    function equalSegments(a: string, b: string) {
        return caseSensitive ? a === b : a.toLowerCase() === b.toLowerCase()
    }

    // Normalize extra slashes.
    if (inputSeparator === '/') {
        from = from.replaceAll(/\/{2,}/g, '/')
        to = to.replaceAll(/\/{2,}/g, '/')
    } else {
        from = from.replaceAll(/\\{2,}/g, '\\')
        to = to.replaceAll(/\\{2,}/g, '\\')
    }

    // Trim trailing slashes.
    if (from !== inputSeparator && from.endsWith(inputSeparator)) {
        from = from.slice(0, -1)
    }
    if (to !== inputSeparator && to.endsWith(inputSeparator)) {
        to = to.slice(0, -1)
    }

    if (equalSegments(from, to)) {
        return ''
    }

    // From a relative to an absolute path, the absolute path dominates.
    if (!from.startsWith(inputSeparator) && to.startsWith(inputSeparator)) {
        return to
    }

    const fromParts = from === inputSeparator ? [''] : from.split(inputSeparator)
    const toParts = to === inputSeparator ? [''] : to.split(inputSeparator)

    // Find the common root.
    let commonLength = 0
    while (
        commonLength < fromParts.length &&
        commonLength < toParts.length &&
        equalSegments(fromParts[commonLength], toParts[commonLength])
    ) {
        commonLength++
    }

    // Calculate the number of directory traversals needed.
    const traversals = fromParts.length - commonLength

    // Build the relative path.
    const relativePath: string[] = []
    for (let i = 0; i < traversals; i++) {
        relativePath.push('..')
    }

    // Add the non-common path parts from the 'to' path.
    relativePath.push(...toParts.slice(commonLength))

    return relativePath.join(outputSeparator)
}

type PathSeparator = '\\' | '/'
