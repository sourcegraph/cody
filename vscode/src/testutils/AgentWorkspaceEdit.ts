import type * as vscode from 'vscode'
import type { Uri, WorkspaceEditEntryMetadata } from 'vscode'
import type { EditFileOperation, WorkspaceEditOperation } from '../jsonrpc/agent-protocol'

export class AgentWorkspaceEdit implements vscode.WorkspaceEdit {
    private edits: WorkspaceEditOperation[] = []

    get operations(): WorkspaceEditOperation[] {
        return Array.from(this.edits.values())
    }

    get size(): number {
        return this.edits.length
    }

    public has(uri: vscode.Uri): boolean {
        const uriString = uri.toString()
        for (const operation of this.edits.values()) {
            switch (operation.type) {
                case 'create-file':
                case 'edit-file':
                    if (operation.uri === uriString) {
                        return true
                    }
                    break
            }
        }
        return false
    }

    public createFile(
        uri: vscode.Uri,
        options?:
            | {
                  readonly overwrite?: boolean | undefined
                  readonly ignoreIfExists?: boolean | undefined
                  readonly contents?: Uint8Array | vscode.DataTransferFile | undefined
              }
            | undefined,
        metadata?: vscode.WorkspaceEditEntryMetadata | undefined
    ): void {
        if (options?.contents && !(options.contents instanceof Uint8Array)) {
            throw new Error(
                `options.contents must be a Uint8Array. Unsupported argument ${options.contents}`
            )
        }
        this.edits.push({
            type: 'create-file',
            uri: uri.toString(),
            options: {
                overwrite: options?.overwrite,
                ignoreIfExists: options?.ignoreIfExists,
            },
            textContents: options?.contents instanceof Uint8Array ? options?.contents?.toString() : '',
            metadata,
        })
    }

    public deleteFile(
        uri: Uri,
        options?: { readonly recursive?: boolean; readonly ignoreIfNotExists?: boolean },
        metadata?: WorkspaceEditEntryMetadata
    ): void {
        throw new Error('Method not implemented.')
    }

    public renameFile(
        oldUri: Uri,
        newUri: Uri,
        options?: { readonly overwrite?: boolean; readonly ignoreIfExists?: boolean },
        metadata?: WorkspaceEditEntryMetadata
    ): void {
        throw new Error('Method not implemented.')
    }

    public replace(
        uri: vscode.Uri,
        range: vscode.Range,
        newText: string,
        metadata?: vscode.WorkspaceEditEntryMetadata
    ): void {
        this.editOperation(uri).edits.push({
            type: 'replace',
            range,
            value: newText,
            metadata,
        })
    }

    public insert(
        uri: vscode.Uri,
        position: vscode.Position,
        content: string,
        metadata?: vscode.WorkspaceEditEntryMetadata
    ): void {
        this.editOperation(uri).edits.push({
            type: 'insert',
            position,
            value: content,
            metadata,
        })
    }

    public delete(
        uri: vscode.Uri,
        range: vscode.Range,
        metadata?: vscode.WorkspaceEditEntryMetadata
    ): void {
        this.editOperation(uri).edits.push({
            type: 'delete',
            range,
            metadata,
        })
    }

    private editOperation(uri: vscode.Uri): EditFileOperation {
        const uriString = uri.toString()
        for (const operation of this.edits.values()) {
            if (operation.type === 'edit-file' && operation.uri === uriString) {
                return operation
            }
        }
        const result: EditFileOperation = {
            type: 'edit-file',
            uri: uri.toString(),
            edits: [],
        }
        this.edits.push(result)
        return result
    }

    // ==================
    // Unimplemented APIs
    // ==================

    public entries(): [vscode.Uri, vscode.TextEdit[]][] {
        throw new Error('Method not implemented.')
    }
    public set(uri: vscode.Uri, edits: readonly (vscode.TextEdit | vscode.SnippetTextEdit)[]): void
    public set(
        uri: vscode.Uri,
        edits: readonly [vscode.TextEdit | vscode.SnippetTextEdit, vscode.WorkspaceEditEntryMetadata][]
    ): void
    public set(uri: vscode.Uri, edits: readonly vscode.NotebookEdit[]): void
    public set(
        uri: vscode.Uri,
        edits: readonly [vscode.NotebookEdit, vscode.WorkspaceEditEntryMetadata][]
    ): void
    public set(uri: unknown, edits: unknown): void {
        throw new Error('Method not implemented.')
    }

    public get(uri: vscode.Uri): vscode.TextEdit[] {
        // Not clear what to do about non-edit operations...
        throw new Error('Method not implemented.')
    }
}
