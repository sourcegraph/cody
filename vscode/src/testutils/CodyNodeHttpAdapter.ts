import type { Request } from '@pollyjs/core'
import HttpAdapter from './pollyjs/NodeHttpAdapter'

export class CodyNodeHttpAdapter extends HttpAdapter {
    public async onRequest(request: Request): Promise<void> {
        if (request.body) {
            request.body = request.body
                .replaceAll(/`([^`]*)(cody-vscode-shim-test[^`]*)`/g, '`$2`')
                .replaceAll(/(\\\\)(\w)/g, '/$2')
        }

        return super.onRequest(request)
    }
}
