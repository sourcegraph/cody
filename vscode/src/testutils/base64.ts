import pako from 'pako'

export function decodeCompressedBase64(text: string): any {
    const bytes = Buffer.from(text, 'base64')
    const unzipped = pako.ungzip(bytes)
    return JSON.parse(Buffer.from(unzipped).toString('utf-8'))
}
