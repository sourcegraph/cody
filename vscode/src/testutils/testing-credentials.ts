import { execSync } from 'node:child_process'
import { DOTCOM_URL } from '@sourcegraph/cody-shared'

export interface TestingCredentials {
    readonly token?: string
    readonly redactedToken?: string
    readonly serverEndpoint: string
}

function loadSecret(name: string): string {
    return execSync(
        `gcloud secrets versions access latest --secret ${name} --project cody-agent-tokens --quiet`
    )
        .toString()
        .trim()
}

export function dotcomCredentials(): TestingCredentials {
    return {
        redactedToken: DOTCOM_TESTING_CREDENTIALS.dotcom.redactedToken,
        serverEndpoint: 'https://sourcegraph.com/',
        token: loadSecret('CODY_PRO_ACCESS_TOKEN'),
    }
}

// See instructions in agent/scripts/export-cody-http-recording-tokens.sh for
// how to update the `redacted` tokens when the access token changes.
export const DOTCOM_TESTING_CREDENTIALS = {
    dotcom: {
        token: process.env.SRC_DOTCOM_PRO_ACCESS_TOKEN,
        redactedToken: 'REDACTED_fc324d3667e841181b0779375f26dedc911d26b303d23b29b1a2d7ee63dc77eb',
        serverEndpoint: DOTCOM_URL.toString(),
    } satisfies TestingCredentials,
    dotcomProUserRateLimited: {
        token: process.env.SRC_DOTCOM_PRO_RATE_LIMIT_ACCESS_TOKEN,
        redactedToken: 'REDACTED_c31e1e5cbed2b06911f09e4e9766c7df227fb23b80cb364c1fe289a845667b4e',
        serverEndpoint: DOTCOM_URL.toString(),
    } satisfies TestingCredentials,
    dotcomUnauthed: {
        token: undefined,
        redactedToken: undefined,
        serverEndpoint: DOTCOM_URL.toString(),
    } satisfies TestingCredentials,
}

export const ENTERPRISE_TESTING_CREDENTIALS = {
    enterprise: {
        token: process.env.SRC_ENTERPRISE_ACCESS_TOKEN,
        redactedToken: 'REDACTED_69e9f79ce29352d014eeb80b56510341844eb82ad9abac7cab3631c7e873e4ce',
        serverEndpoint: 'https://demo.sourcegraph.com/',
    } satisfies TestingCredentials,
    s2: {
        token: process.env.SRC_S2_ACCESS_TOKEN,
        redactedToken: 'REDACTED_1858aad0e1ff07ae26d4042086acb9da455866ad617afd2cb9ab9419e1be1104',
        serverEndpoint: 'https://sourcegraph.sourcegraph.com/',
    } satisfies TestingCredentials,
    s2Unauthed: {
        token: undefined,
        redactedToken: undefined,
        serverEndpoint: 'https://sourcegraph.sourcegraph.com/',
    } satisfies TestingCredentials,
}

export const TESTING_CREDENTIALS: typeof ENTERPRISE_TESTING_CREDENTIALS &
    typeof DOTCOM_TESTING_CREDENTIALS = {
    ...DOTCOM_TESTING_CREDENTIALS,
    ...ENTERPRISE_TESTING_CREDENTIALS,
}
