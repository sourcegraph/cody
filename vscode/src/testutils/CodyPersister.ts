// TODO: This potentially introduces some breaking behaviour compared to the previous one.
// When the playwright E2E lands properly we should re-record all agent tests
// with this updated recorder.

import crypto from 'node:crypto'

import type { Har, HarEntry } from '@pollyjs/persister'
import FSPersister from '@pollyjs/persister-fs'
import { type CompletionData, isError, parseEvents } from '@sourcegraph/cody-shared'
import { CompletionsResponseBuilder } from '@sourcegraph/cody-shared/src/sourcegraph-api/completions/CompletionsResponseBuilder'
import { PollyYamlWriter } from './PollyYamlWriter'
import { decodeCompressedBase64 } from './base64'

const AUTH_HEADER_REGEX = /^(?<prefix>token|bearer)\s+(?<redacted>REDACTED_)?(?<token>.*?)\s*$/im
/**
 * SHA-256 digests a Sourcegraph access token so that it's value is redacted but
 * remains uniquely identifyable. The token needs to be uniquely identifiable so
 * that we can correctly replay HTTP responses based on the access token.
 */
export function redactAuthorizationHeader(header: string): string {
    const match = AUTH_HEADER_REGEX.exec(header)
    if (match) {
        if (match.groups?.redacted) {
            return header
        }
        const tokenHash = sha256(`prefix${match.groups?.token}`)
        return `${match.groups?.prefix} REDACTED_${tokenHash}`
    }
    return header
}

function sha256(input: string): string {
    return crypto.createHash('sha256').update(input).digest('hex')
}

/**
 * The default file system persister with the following customizations
 *
 * - Replaces Cody access tokens with the string "REDACTED" because we don't
 *   want to commit the access token into git.
 * - To avoid diff churn/conflicts:
 *   - Sets date headers to a known static date
 *   - Removes cookies
 *   - Sets dates/timing information stored by Polly to static values
 */
export class CodyPersister extends FSPersister {
    // HACK: `FSPersister` has a private `api` property that writes the
    // recording.har file using JSON format. We override the `api` property here
    // with a custom implementation that uses YAML format instead. This property
    // is intentionally marked as public even if it's not used anywhere.
    public api: PollyYamlWriter

    constructor(polly: any) {
        super(polly)
        if (!this.options.recordingsDir) {
            throw new Error('No recording directory provided')
        }
        this.api = new PollyYamlWriter(this.options.recordingsDir)
    }
    public static get id(): string {
        return 'fs'
    }

    public async onFindRecording(recordingId: string): Promise<Har | null> {
        const har = await super.onFindRecording(recordingId)
        if (har === null) {
            return har
        }
        for (const entry of har.log.entries) {
            const postData = entry?.request?.postData
            if (
                postData !== undefined &&
                postData?.text === undefined &&
                (postData as any)?.textJSON !== undefined
            ) {
                // Format `postData.textJSON` back into the escaped string for the `.text` format.
                postData.text = JSON.stringify((postData as any).textJSON)
                ;(postData as any).textJSON = undefined
            }
        }
        return har
    }

    public onSaveRecording(recordingId: string, recording: Har): Promise<void> {
        const entries = recording.log.entries
        recording.log.entries.sort((a, b) => a.request.url.localeCompare(b.request.url))
        for (const entry of entries) {
            if (entry.request?.postData?.text?.startsWith('{')) {
                // Format `postData.text` as a JSON object instead of escaped string.
                // This makes it much easier to review the har file locally.
                const postData: any = entry.request.postData
                postData.textJSON = JSON.parse(entry.request.postData.text)
                postData.text = undefined
            }
            // Clean up the entries to reduce the size of the diff when re-recording
            // and to remove any access tokens.
            const headers = [...entry.request.headers, ...entry.response.headers]
            for (const header of headers) {
                switch (header.name.toLowerCase()) {
                    case 'authorization':
                        header.value = redactAuthorizationHeader(header.value)
                        break
                    // We should not harcode the dates to minimize diffs because
                    // that breaks the expiration feature in Polly.
                }
            }

            // Remove any headers and cookies we don't need at all.
            entry.request.headers = this.filterHeaders(entry.request.headers)
            entry.response.headers = this.filterHeaders(entry.response.headers)
            entry.response.content.text
            entry.request.cookies.length = 0
            entry.response.cookies.length = 0
            entry.response.content.text = this.postProcessHarEntryResponse(entry)

            // Compared to V1 we don't nullify time fields as instead they can be configured on
            // playback with adjustable speed. This makes it much easier to play
            // back a test at the original recorded speed.

            // entry.time = 0
            // entry.timings = {}

            const responseContent = entry.response.content
            if (
                responseContent?.encoding === 'base64' &&
                responseContent?.mimeType === 'application/json' &&
                responseContent.text
            ) {
                // The GraphQL responses are base64+gzip encoded. We decode them
                // in a sibling `textDecoded` property so we can more easily review
                // in in pull requests.
                try {
                    const text = JSON.parse(responseContent.text)[0]
                    const decodedBase64 = decodeCompressedBase64(text)
                    ;(responseContent as any).textDecoded = decodedBase64
                } catch {
                    // Ignored: uncomment below to debug. It's fine to ignore this error because we only
                    // make a best-effort to decode the gzip+base64 encoded JSON payload. It's not needed
                    // for the HTTP replay to work correctly because we leave the `.text` property unchanged.
                    // console.error('base64 decode error', error)
                }
            }
        }
        return super.onSaveRecording(recordingId, recording)
    }

    private postProcessHarEntryResponse(entry: HarEntry): string | undefined {
        const { text } = entry.response.content
        if (text === undefined) {
            return undefined
        }
        if (
            !entry.request.url.includes('/.api/completions/stream') &&
            !entry.request.url.includes('/completions/code')
        ) {
            return text
        }
        return postProcessCompletionsStreamText(entry.request.url, text)
    }

    private filterHeaders(
        headers: { name: string; value: string }[]
    ): { name: string; value: string }[] {
        const removeHeaderNames = new Set([
            'set-cookie',
            'server',
            'via',
            'x-sourcegraph-actor-anonymous-uid',
            //TODO(rnauta): leaky abstraction, how to configure this on the other end
            'x-mitm-proxy-endpoint',
            'x-mitm-auth-available',
        ])
        const removeHeaderPrefixes = ['x-trace', 'cf-']
        return headers.filter(
            header =>
                !removeHeaderNames.has(header.name) &&
                removeHeaderPrefixes.every(prefix => !header.name.startsWith(prefix))
        )
    }
}

export function postProcessCompletionsStreamText(url: string, text: string): string | undefined {
    const builder = CompletionsResponseBuilder.fromUrl(url)
    const parseResult = parseEvents(builder, text)
    if (isError(parseResult)) {
        return text
    }
    const hasError = parseResult.events.some(event => event.type === 'error')
    if (hasError) {
        return text
    }

    const [completionEvent, doneEvent] = parseResult.events.slice(-2)
    if (completionEvent?.type !== 'completion' || doneEvent?.type !== 'done') {
        return text
    }

    const data: CompletionData =
        builder.apiVersion >= 2
            ? {
                  deltaText: completionEvent.completion,
                  stopReason: completionEvent.stopReason,
              }
            : {
                  completion: builder.totalCompletion,
                  stopReason: completionEvent.stopReason,
              }

    // NOTE(olafurpg) after api-version=2, we send delta text on SSE events
    // instead of the full completion including prefix. This was a huge performance
    // win but it's impractical for the HTTP record/replay format where we want
    // to generate as few lines as possible. For this reason, we post-process
    // the api-version=2 output to make it look like api-version=1 where the
    // full completion is included via `.completion`.
    const result = `event: completion
data: ${JSON.stringify(data)}

event: done
data: {}

`
    return result
}
