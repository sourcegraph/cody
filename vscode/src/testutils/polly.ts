import { execSync } from 'node:child_process'
import path from 'node:path'
import jsonStableStringify from 'fast-json-stable-stringify'

import { type EXPIRY_STRATEGY, type Headers, type MODE, Polly, type PollyConfig } from '@pollyjs/core'

import { CodyNodeHttpAdapter } from './CodyNodeHttpAdapter'
import { CodyPersister, redactAuthorizationHeader } from './CodyPersister'

interface PollyOptions {
    recordingName: string
    keepUnusedRecordings?: boolean
    recordingDirectory?: string
    recordIfMissing?: boolean
    recordingMode?: MODE
    recordingExpiryStrategy?: EXPIRY_STRATEGY
    expiresIn?: string | null
}

export function startPollyRecording(userOptions: PollyOptions): Polly {
    const options = defaultPollyOptions(userOptions)
    Polly.register(CodyNodeHttpAdapter)
    Polly.register(CodyPersister)
    return new Polly(options.recordingName ?? 'CodyAgent', {
        flushRequestsOnStop: true,
        recordIfMissing: options.recordIfMissing ?? options.recordingMode === 'record',
        mode: options.recordingMode,
        adapters: [CodyNodeHttpAdapter],
        persister: CodyPersister,
        recordFailedRequests: true,
        expiryStrategy: options.recordingExpiryStrategy,
        expiresIn: options.expiresIn,
        persisterOptions: {
            keepUnusedRequests: options.keepUnusedRecordings ?? true,
            fs: {
                recordingsDir: options.recordingDirectory,
            },
        },
        matchRequestsBy: defaultMatchRequestsBy,
    })
}

const defaultMatchRequestsBy: PollyConfig['matchRequestsBy'] = {
    order: false,

    // Canonicalize JSON bodies so that we can replay the recording even if the JSON strings
    // differ by semantically meaningless things like object key enumeration order.
    body(body) {
        try {
            if (typeof body === 'string' && (body.startsWith('{') || body.startsWith('['))) {
                return jsonStableStringify(JSON.parse(body))
            }
        } catch {}
        return body
    },

    // The logic below is a bit tricky to follow. Simplified, we need to
    // ensure that Polly generates the same request ID regardless if
    // we're running in record mode (with an access token) or in replay
    // mode (with a redacted token). The ID is computed by Polly as the
    // MD5 digest of all request "identifiers", which a JSON object that
    // includes a "headers" property from the result of the function
    // below. To better understand what's going on, it's helpful to read
    // the implementation of Polly here:
    //   https://sourcegraph.com/github.com/Netflix/pollyjs@9b6bede12b7ee998472b8883c9dd01e2159e00a8/-/blob/packages/@pollyjs/core/src/-private/request.js?L281
    headers(headers): Headers {
        // Get the authorization token.
        const { authorization } = headers
        let header =
            typeof authorization === 'string'
                ? authorization
                : Array.isArray(authorization)
                  ? authorization.at(0)
                  : undefined

        // Redact it so that the ID is the same regardless if we're in record or replay
        // mode.
        if (header) {
            header = redactAuthorizationHeader(header)
        }

        // Normalize to always be a single header value (not an array).
        return header ? { authorization: header } : {}
    },
}

function defaultPollyOptions(
    options: Pick<PollyOptions, 'recordingName' | 'recordingDirectory'>
): PollyOptions {
    let recordingMode: MODE = 'replay'
    switch (process.env.CODY_RECORDING_MODE) {
        case 'record':
        case 'replay':
        case 'passthrough':
        case 'stopped':
            recordingMode = process.env.CODY_RECORDING_MODE
            break
        default:
            if (typeof process.env.CODY_RECORDING_MODE === 'string') {
                throw new TypeError(
                    `Not a valid recording mode '${process.env.CODY_RECORDING_MODE}'. Valid options are record, replay, passthrough, or stopped.`
                )
            }
    }
    const recordingDirectory = (): string => {
        const rootDirectory = execSync('git rev-parse --show-toplevel', {
            encoding: 'utf-8',
        }).trim()
        return path.join(rootDirectory, 'recordings')
    }
    return {
        recordIfMissing: process.env.CODY_RECORD_IF_MISSING === 'true' || recordingMode === 'record',
        recordingMode,
        recordingDirectory: options.recordingDirectory ?? recordingDirectory(),
        expiresIn: '365d',
        recordingExpiryStrategy: 'error',
        ...options,
    }
}
