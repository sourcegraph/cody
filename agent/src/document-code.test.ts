import path from 'node:path'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import { TESTING_CREDENTIALS } from '../../vscode/src/testutils/testing-credentials'
import { TestClient } from './TestClient'
import { TestWorkspace } from './TestWorkspace'

describe('Document Code', { timeout: 5000 }, () => {
    const workspace = new TestWorkspace(path.join(__dirname, '__tests__', 'document-code'))
    const client = TestClient.create({
        workspaceRootUri: workspace.rootUri,
        name: 'document-code',
        credentials: TESTING_CREDENTIALS.dotcom,
    })

    beforeAll(async () => {
        await workspace.beforeAll()
        await client.beforeAll()
    })

    afterAll(async () => {
        await workspace.afterAll()
        await client.afterAll()
    })

    it('editCommands/document (basic function)', async () => {
        expect(await client.documentCode(workspace.file('src', 'sum.ts'))).toMatchSnapshot()
    })

    it('commands/document (Method as part of a class)', async () => {
        expect(await client.documentCode(workspace.file('src', 'TestClass.ts'))).toMatchSnapshot()

        const { requests } = await client.request('testing/networkRequests', null)
        expect(requests.find(r => r.body?.includes('const longSuffix'))).toBeTruthy()
        expect(requests.find(r => r.body?.includes('const longPrefix'))).toBeTruthy()
    })

    it('commands/document (Function within a property)', async () => {
        expect(await client.documentCode(workspace.file('src', 'TestLogger.ts'))).toMatchSnapshot()
    })

    it('commands/document (nested test case)', async () => {
        expect(await client.documentCode(workspace.file('src', 'example.test.ts'))).toMatchSnapshot()
    })

    it('commands/document (Kotlin class name)', async () => {
        expect(await client.documentCode(workspace.file('src', 'Hello.kt'))).toMatchSnapshot()
    })
})
