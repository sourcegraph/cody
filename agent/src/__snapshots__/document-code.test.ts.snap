// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Document Code > commands/document (Function within a property) 1`] = `
"const foo = 42
export const TestLogger = {
    startLogging: () => {
        // Do some stuff

        /**
         * Records a log message to the console.
         */
        function recordLog() {
            console.log(/* CURSOR */ 'Recording the log')
        }

        recordLog()
    },
}
"
`;

exports[`Document Code > commands/document (Kotlin class name) 1`] = `
"/**
 * Provides a greeting message.
 * @return The greeting message "Hello, world!"
 */
class He/* CURSOR */llo {
    fun greeting(): String {
        return "Hello, world!"
    }
}
"
`;

exports[`Document Code > commands/document (Method as part of a class) 1`] = `
"const foo = 42
// Should be present in the LLM prompt
const longPrefix = \`
    longPrefix content
    longPrefix content
    longPrefix content
    longPrefix content
    longPrefix content
    longPrefix content
    longPrefix content
\`

export class TestClass {
    constructor(private shouldGreet: boolean) {}

    /**
     * Logs "Hello World!" to the console if the \`shouldGreet\` property is true.
     */
    public functionName() {
        if (this.shouldGreet) {
            console.log(/* CURSOR */ 'Hello World!')
        }
    }
}

// Should be present in the LLM prompt
const longSuffix = \`
    longSuffix content
    longSuffix content
    longSuffix content
    longSuffix content
    longSuffix content
    longSuffix content
    longSuffix content
\`
"
`;

exports[`Document Code > commands/document (nested test case) 1`] = `
"import { expect } from 'vitest'
import { it } from 'vitest'
import { describe } from 'vitest'

describe('test block', () => {
    it('does 1', () => {
        expect(true).toBe(true)
    })

    it('does 2', () => {
        expect(true).toBe(true)
    })

    it('does something else', () => {
        // This line will error due to incorrect usage of \`performance.now\`
        /**
         * Retrieves the current time in milliseconds since the page was loaded, for use in performance measurements.
         */
        const startTime = performance.now(/* CURSOR */)
    })
})
"
`;

exports[`Document Code > editCommands/document (basic function) 1`] = `
"/**
 * Adds two numbers together and returns the result.
 * 
 * @param a - The first number to add.
 * @param b - The second number to add.
 * @returns The sum of the two input numbers.
 */
export function sum(a: number, b: number): number {
    /* CURSOR */
}
"
`;
